text/0000-elementary-functions.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: elementary function calculation
+- Start Date: 2019-02-11
+- RFC PR: (leave this empty)
+- Authors:
+  - Matthias Güdemann (matthias.gudemann@iohk.io)
+
+# Summary
+[summary]: #summary
+
+The delegation mechanism uses a decay function for refunds which involves
+calculating the exponential function. The moving average of the performance of a
+stake pool is calculated using exponentiation of two fractional numbers.
+
+This proposal addresses the method of how to calculate the exponential function,
+the natural logarithm and the exponentiation using a non-integer exponent.
+
+# Motivation
+[motivation]: #motivation
+
+The delegation requires some non-integral calculations using elementary
+functions. These calculations have to be approximated and are not standardized
+in general.
+
+In order to prevent forks because of different results on different
+implementations or architectures, we provide our own implementation of the
+necessary approximation functions.
+
+Assuming the same representation of non-integral values and standardized basic
+arithmetic operations, this leads to equivalent results using different
+implementation languages or CPU architectures.
+
+# Decision
+[Decision]: #decision
+
+The library `non-integer` exports the functions `exp'` which approximates the
+exponential function, `ln'` which approximates the natural logarithm and the
+infix operator `(***)` which approximates exponentiation.
+
+Current implementation: `dependencies/non-integer` of `fm-ledger-rules@727cca6`
+(cf. submodule pointer).
+
+All input parameters for the functions need to be instances of `RealFrac` in
+order to support addition, subtraction, multiplication and division. The
+approximation algorithm terminates if the basic operations terminate. The
+approximation is recursive, the recursion depth is bounded by both a fixed
+constant and the precision of succeeding incremental approximations.
```

Comment:
This is really cute. But this is haskell specific. We need to provide a concrete decision that can be implemented and be realistic on other languages than haskell too. Unlike the formal spec, here we actually need to be very precise about what we are proposing.

Here you are talking about recursion. Even bounded by a parameter, this is an unacceptable solution as we need to be able to run this in limited/constrained environment too.

Reply:
recursion is just an iteration method here, loops can be used just as well